compiler/rustc_codegen_ssa/src/mir/intrinsic.rs

```diff
@@ -581,10 +581,10 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
 fn int_type_width_signed(ty: Ty<'_>, tcx: TyCtxt<'_>) -> Option<(u64, bool)> {
     match ty.kind() {
         ty::Int(t) => {
-            Some((t.bit_width().unwrap_or(u64::from(tcx.sess.target.pointer_width)), true))
+            Some((t.bit_width().unwrap_or_else(|| u64::from(tcx.sess.target.pointer_width)), true))
```

Comment:
This makes the code less readable and does not improve performance

Reply:
The readability is, in my view, minimally affected. 
Performance wise, is the `u64::from` that is being called there inlined or not?

Reply:
yes, if not that's a bug, the function is trivial

Reply:
If the closure gets inlined (trivial body, no `dyn`) and the `u64::from` call stays inlined, the only trade off is a minimal loss of readability in exchange for no clippy warnings.

Reply:
This applies to many of the conversions done here, not just this one. I'm not a fan of making code worse in order to make Clippy shut up. If linters get in the way like this *they* need to be fixed.

Reply:
I guess it is time to sprinkle `#[allow(clippy::or_fun_call)]` in a few places.

Reply:
I personally also feel that some of these changes slightly worsen readability without having a benefit apart from silencing clippy. I do agree with https://github.com/rust-lang/rust/issues/78814#issuecomment-723257685 here and would prefer to explicitly uplift lints we deem useful, or even keep them in clippy and add them to CI if they are really relevant but do not quite fit into rustc, for example `invalid_regex`.

Reply:
> The option to get clippy out of the way is to sprinkle #[allow(clippy::or_fun_call)] in the few places where this is deemed overly detrimental.

That is assuming the *goal* is to reduce clippy lints. But the lints are only a means to an end, especially when they are inaccurate heuristics. The actual goal of the lint in question is to improve performance. If there are many cases when it doesn't it should only be used as a tool to search for optimization candidates, not as a guide how to rewrite all code.

Reply:
> > The option to get clippy out of the way is to sprinkle #[allow(clippy::or_fun_call)] in the few places where this is deemed overly detrimental.
> 
> That is assuming the _goal_ is to reduce clippy lints. But the lints are only a means to an end, especially when they are inaccurate heuristics. The actual goal of the lint in question is to improve performance. If there are many cases when it doesn't it should only be used as a tool to search for optimization candidates, not as a guide how to rewrite all code.

The goal of this *specific pull request* is to reduce lints. While I may agree that some lints are inaccurate and tend towards false positives and specific workarounds, making pull requests like this one seems to me to be the only way to determine with a high degree of certainty which lints should be accepted.

Reply:
> The goal of this specific pull request is to reduce lints.

Well yes, it's clear that's why you made the PR. The question is whether that's a worthwhile goal, and whether silencing the lints is worth the additional noise in the codebase.

> making pull requests like this one seems to me to be the only way to determine with a high degree of certainty which lints should be accepted.

I'm not sure what you mean here - are you saying that we don't know which lints are worthwhile unless there's some discussion about it? There has been discussion about it, and most people seem to agree this change isn't worth it in all cases.

Reply:
> I'm not sure what you mean here - are you saying that we don't know which lints are worthwhile unless there's some discussion about it? There has been discussion about it, and most people seem to agree this change isn't worth it in all cases.

So then the question should be "which subset of cases are most useful?". For example, some changes were like those in `compiler/rustc_ast/src/attr/mod.rs` lines 69, 142, and 186, where changing `unwrap_or` to `unwrap_or_else` involved only dropping the final empty parenthesis. Some changes were to `unwrap_or_default`, which to me offers both better readability and better performance.

Reply:
Ok, I see. I think we can make the improvement you're talking about without also adding the blanket `allow`, which just adds noise. I don't think getting rustc to 0 clippy warnings should be a goal.

Reply:
> Ok, I see. I think we can make the improvement you're talking about without also adding the blanket `allow`, which just adds noise. I don't think getting rustc to 0 clippy warnings should be a goal.

You are suggesting a basic floor of "ignored noisy lints" that would be incorporated at sometime into the CI pipeline if clippy ever becomes part of it. There may be some technical difficulties that I may not see in that suggestion, but that is something which I would be fine with, as long as that floor is kept sufficiently low.